### PR TITLE
licensecheck: define new Type type

### PR DESCRIPTION
**Pull request by rsc**
The new Type type is a bit set of license properties
instead of trying to define a fixed set of license categories.

Also set up test of license types.
Only add one license type, WTFPL ⇒ Discouraged, for testing.

### CODE REVIEW

File: license.go [hunk 1]
```diff
@@ -132,3 +107,112 @@ type Match struct {
...
+// For example, (ShareServer|NonCommercial).String() = "ShareServer|NonCommercial".
+func (t Type) String() string {
+	if t == 0 {
+		return "Unrestricted"
+	}
+	s := ""
+	for _, b := range typeBits {
+		if b.t != 0 && t&b.t == b.t {
+			t &^= b.t
+			s += "|" + b.s
```
Comment (robpike) [inline license.go:185]: doesn't this mean the string always starts with a "|"? that's not what you want, is it?

Comment (rsc) [reply license.go:185]: It's easier to build that way and gets sliced away at the bottom.

**Comment by rsc** [comment]
I think the idea of Type being a set of properties is still a good one,
but I'm bothered by Unknown not being the zero of Type.
You have to go out of your way to make sure to set Type to Unknown,
and if you forget you get Unrestricted instead, which is very wrong.

An alternative would be to go back to plain bits with Unknown = 0,
giving up on bit operations like OR being meaningful.

	Unknown = 0
	Unrestricted = 1<<0
	Notice = 1<<1
	ShareLibrary = 1<<2 
	ShareProgram = 1<<3
	ShareServer = 1<<4
	NonCommercial = 1<<5
	Discouraged = 1<<6

Instead I think we'd have to define

	// Merge returns the overall license type of a program
	// combining programs of license types t and u.
	// If t.Merge(u) != t, then u is a more restrictve license type than t.
	func (t Type) Merge(u Type) Type {
		// Unrestricted...ShareServer are mutually exclusive
		// and have increasing levels of restrictions.
		// Take the max.
		const M = Unrestricted|Notice|ShareLibrary|ShareProgram|ShareServer
		tm, t := t&M, t&^M
		um, u := u&M, u&^M
		if tm < um {
			tm = um
		}
		if tm&(tm-1) != 0 {
			// Not a power of two. Save only the high bit.
			for bit := ShareServer; bit > 0; bit >>= 1 {
				if tm&bit != 0 {
					tm = bit
					break
				}
			}
		}
		
		// OR the other bits together.
		return t | u | tm
	}

And then to look for conflicts you'd look for t.Merge(u) != t where t is the top-level package's license type and u is the license type of a dependency.

Thoughts?

**Comment by rsc** [comment]
I made the change to arrange for the zero value of Type to be Unknown, as it should be.

**Pull request merged**